test/unit/puppetlabs/jruby_utils/lockable_ref_pool_test.clj

```diff
@@ -0,0 +1,638 @@
+                (str "timed out waiting for the borrow after lock requested "
+                     "thread to finish"))
+            (is (true? (timed-deref
+                        borrow-after-lock-acquired-instance-acquired?))
+                "timed out waiting for the borrow after lock acquired")
+            (is (true? (timed-deref
+                        borrow-after-lock-acquired-thread))
+                (str "timed out waiting for the borrow after lock acquired "
+                     "thread to finish")))))
+            (is (not (.isLocked pool))))))
+
+(deftest pool-lock-supersedes-existing-borrows-test
+  (testing "if there are pending borrows when pool.lock() is called, they aren't fulfilled until after unlock()"
+    (let [pool (create-populated-pool 2)
+          instances (borrow-n-instances pool 2)
+          blocked-borrow-thread-started? (promise)
+          blocked-borrow-thread-borrowed? (promise)
+          blocked-borrow-thread (future (deliver blocked-borrow-thread-started? true)
+                                        (let [instance (.borrowItem pool)]
+                                          (deliver blocked-borrow-thread-borrowed? true)
+                                          (.releaseItem pool instance))
+                                        true)
+          lock-thread-started? (promise)
+          lock-acquired? (promise)
+          unlock? (promise)
+          lock-thread (future (deliver lock-thread-started? true)
+                              (.lock pool)
+                              (deliver lock-acquired? true)
+                              @unlock?
+                              (.unlock pool)
+                              true)]
+      @blocked-borrow-thread-started?
+      @lock-thread-started?
+      (is (not (realized? blocked-borrow-thread-borrowed?)))
+      (let [start (System/currentTimeMillis)]
+        (while (and (not (.isLocked pool))
+                    (< (- (System/currentTimeMillis) start) 10000))
+          (Thread/yield)))
+      (is (.isLocked pool))
+      (is (not (realized? lock-acquired?)))
+
+      (return-instances pool instances)
+      (is (not (realized? blocked-borrow-thread-borrowed?)))
+      (is (not (realized? lock-thread)))
+      (is (true? (timed-deref lock-acquired?))
+          "timed out waiting for the lock to be acquired")
+      (is (.isLocked pool))
+
+      (deliver unlock? true)
+      (is (true? (timed-deref lock-thread))
+          "timed out waiting for the lock thread to finish")
+      (is (true? (timed-deref blocked-borrow-thread-borrowed?))
+          "timed out waiting for the borrow to complete")
+      (is (true? (timed-deref blocked-borrow-thread))
+          "timed out waiting for the borrow thread to complete")
+      (is (not (.isLocked pool))))))
+
+(deftest pool-lock-reentrant-for-borrow-from-locking-thread
+  (testing "the thread that holds the pool lock may borrow instances while holding the lock"
```

Comment:
Why do we want to allow this?

Reply:
I actually make use of it several times in the ReferencePool implementation of the Clojure protocol. It's okay because the thread that locks the pool does so in order to have complete control over all of its instances, so it can do things like modify and delete them. Since the only way to get a reference to an instance is by borrowing it, we need to allow borrows to happen by the locking thread while the pool is locked.

A great example of this is when we are deleting an instance. One part of deleting an instance is to clean up the Ruby scripting container associated with it. For the ReferencePool in particular, where the instance can be borrowed by lots of threads, we need lock the pool in order to make sure no one else is borrowing the instance, then get a reference to it (by borrowing it ourselves) and clean up its scripting container, then return our reference back to the pool. See https://github.com/puppetlabs/jruby-utils/pull/78/files#diff-1392e2eb7a0c370666b5d0ec72185b55R66-R74 for an example. Note that we actually return before we clean up here. That's both safe -- because we've prevented anyone else from getting ahold of it because we've locked the pool --  and required, because another thing `cleanup` does is unregister the instance, which we only allow if the pool's current borrow count is 0. I don't particularly _like_ this logic, but it's kind of the only option given that the cleanup code is in Clojure. (Maybe we could re-evaluate that? I think Justin was musing about the possibility at some point)